lib/types.nix

```diff
@@ -160,118 +163,180 @@ let
+            # https://github.com/NixOS/nixpkgs/pull/173568 and
+            # https://github.com/NixOS/nixpkgs/pull/168295 that attempted to revert this,
+            # https://github.com/NixOS/nixpkgs/issues/191124 and
+            # https://github.com/NixOS/nixos-search/issues/391 for what happens if you ignore
+            # this disclaimer.
+            check ? (x: true),
+            # Merge a list of definitions together into a single value.
+            # This function is called with two arguments: the location of
+            # the option in the configuration as a list of strings
+            # (e.g. ["boot" "loader "grub" "enable"]), and a list of
+            # definition values and locations (e.g. [ { file = "/foo.nix";
+            # value = 1; } { file = "/bar.nix"; value = 2 } ]).
+            merge ? mergeDefaultOption,
+            # Whether this type has a value representing nothingness. If it does,
+            # this should be a value of the form { value = <the nothing value>; }
+            # If it doesn't, this should be {}
+            # This may be used when a value is required for `mkIf false`. This allows the extra laziness in e.g. `lazyAttrsOf`.
+            emptyValue ? { },
+            # Return a flat attrset of sub-options. Used to generate
+            # documentation.
+            getSubOptions ? prefix: { },
+            # List of modules if any, or null if none.
+            getSubModules ? null,
+            # Function for building the same option type with a different list of
+            # modules.
+            substSubModules ? m: null,
+            # Function that merge type declarations.
+            # internal, takes a functor as argument and returns the merged type.
+            # returning null means the type is not mergeable
+            typeMerge ? defaultTypeMerge functor,
+            # The type functor.
+            # internal, representation of the type as an attribute set.
+            #   name: name of the type.
+            #   type: function to initialize the type with the passed values.
+            #   wrapped: the type wrapped in case of compound types.
+            #   payload: values passed to the function that initializes the type,
+            #            two payloads of the same type must be combinable with
+            #            the binOp binary operation.
+            #   binOp: binary operation that merge two payloads of the same type.
+            functor ? defaultFunctor self,
+            # The deprecation message to display when this type is used by an option
+            # If null, the type isn't deprecated
+            deprecationMessage ? null,
+            # The types that occur in the definition of this type. This is used to
+            # issue deprecation warnings recursively. Can also be used to reuse
+            # nested types
+            nestedTypes ? { },
+            __uncall__,
+            ...
```

Comment:
We don't accept arbitrary arguments with mkOptionType do you propose we need to change this?